x/shield/keeper/purchase.go

```diff
@@ -98,21 +99,31 @@ func (k Keeper) purchaseShield(ctx sdk.Context, poolID uint64, shield sdk.Coins,
 
 	// Check pool shield limit.
 	poolParams := k.GetPoolParams(ctx)
+	protectionEndTime := ctx.BlockTime().Add(poolParams.ProtectionPeriod)
 	maxShield := sdk.MinInt(pool.ShieldLimit, totalCollateral.Sub(totalWithdrawing).ToDec().Mul(poolParams.PoolShieldLimit).TruncateInt())
 	if shieldAmt.Add(pool.Shield).GT(maxShield) {
 		return types.Purchase{}, types.ErrPoolShieldExceedsLimit
 	}
 
-	// Send service fees to the shield module account and update service fees.
-	if err := k.DepositNativeServiceFees(ctx, serviceFees, purchaser); err != nil {
-		return types.Purchase{}, err
+	// get next purchase ID and set purchase ID after that
+	purchaseID := k.GetNextPurchaseID(ctx)
+	k.SetNextPurchaseID(ctx, purchaseID+1)
+	if !staking {
+		// Send service fees to the shield module account and update service fees.
+		if err := k.DepositNativeServiceFees(ctx, serviceFees, purchaser); err != nil {
+			return types.Purchase{}, err
+		}
+		totalServiceFees := k.GetServiceFees(ctx)
+		totalServiceFees = totalServiceFees.Add(types.MixedDecCoins{Native: sdk.NewDecCoinsFromCoins(serviceFees...)})
+		k.SetServiceFees(ctx, totalServiceFees)
+		totalRemainingServiceFees := k.GetRemainingServiceFees(ctx)
+		totalRemainingServiceFees = totalRemainingServiceFees.Add(types.MixedDecCoins{Native: sdk.NewDecCoinsFromCoins(serviceFees...)})
+		k.SetRemainingServiceFees(ctx, totalRemainingServiceFees)
+	} else {
+		if err := k.AddStaking(ctx, poolID, purchaser, purchaseID, serviceFees.AmountOf(bondDenom), protectionEndTime); err != nil {
```

Comment:
If it is a staking purchase, `serviceFees` will always be zero, right?

Reply:
I just pass in the staking amount as serviceFees. Probably not the best idea 🤔 will fix